Fix outgoing qos2 message id bug (#1045)

larshesel opened the pull request:
Fixes #1045 

@dergraf note my assumption about `pub_rel`s being send to the session before any normal publishes (without message ids) - I hope this is correct.

dergraf commented on apps/vmq_server/src/vmq_mqtt5_fsm.erl:
```diff
@@ -1376,11 +1376,19 @@ handle_messages([{deliver, QoS, Msg} = Obj|Rest], Frames, PubCnt, State, Waiting
+    %% The pubrel messages has fixed message-ids and we therefore need
+    %% ensure the first message id assigned to a publish message is
+    %% larger than any of the pubrel messages to avoid conflicts.
+    NextMsgId = lists:max([MsgId + 1, CurrNextMsgId]),
```
The `MsgId + 1` could theoretically overrun the message id limit `max(ushort)`.

larshesel replied:
Good point.. Will fix that.

dergraf commented on apps/vmq_server/test/vmq_publish_SUITE.erl:
```diff
@@ -459,6 +460,87 @@ publish_c2b_retry_qos2_test(_Config) ->
     ok = expect_alive(Socket),
     ok = gen_tcp:close(Socket).
 
+publish_b2c_ensure_valid_msg_ids_test(Config) ->
```
it would be interesting to test what happens if multiple of such PUBRELs were persisted.

larshesel replied:
Which test scenario are you thinking about here? I agree more cases could/should be covered, ideally we should property test stuff like this, if possible.

larshesel commented:
@dergraf I reverted the incorrect fix and implemented a new one where we store the `next_msg_id` in the queue. This, however, revealed another problem (the `in_order_delivery_SUITE` started failing): namely that we do not remember the `msg_id` on outgoing qos2 messages (which didn't yet receive a `pubrec`) when we store waiting acks - so when the client comes back the outgoing publish will set the dup flag to ture and assign a new `msg_id` which is against the spec. So next step is to also remember the msgid used when sending the initial qos2 publish message out when storing the unacked stuff in the queue.

dergraf commented on apps/vmq_server/src/vmq_mqtt_fsm.erl:
```diff
@@ -539,18 +539,20 @@ check_user(#mqtt_connect{username=User, password=Password} = F, State) ->
                     case vmq_reg:register_subscriber(CAPSettings#cap_settings.allow_register, CoordinateRegs, SubscriberId, CleanSession, QueueOpts) of
-                        {ok, SessionPresent, QPid} ->
+                        {ok, #{session_present := SessionPresent,
+                               initial_msg_id := MsgId}, QPid} ->
```
We could return the `QPid` also as part of the map.

larshesel replied:
yeah, I guess that makes sense. Then spec it as a mandatory field in the map.

larshesel commented:
Since last I:

 - moved the queue pid into the map returned by `vmq_reg:register_subscriber/4`
 - changed the `{deliver, QoS, Msg}` tuple into a record `#deliver{msg_id, qos, msg}` and use this everywhere *except* when enqueueing as we do this across nodes. For some scenarious this means I have to translate between the two, for example in `vmq_queue:enqueue_many` and when draining.

larshesel commented:
Had to rebase to get CI to run...

larshesel commented:
@boboxiaodd, I think this is about done - would you be able to test it out?

boboxiaodd commented:
@larshesel , Great, i clone the master  and  test it now!
